### PR TITLE
[Fixes #7658] FileStreamResultExecutor fails to Dispose FileStream

### LINKED ISSUE #7658: FileStreamResultExecutor fails to Dispose FileStream
### Functional impact
Version 2.0 of ASP.NET Core and onwards appears to fail to call `Dispose()` in the `FileStreamResultExecutor` class under certain circumstances e.g. handing HTTP HEAD requests. This behaviour worked correctly in Version 1.1.

For example the following execution paths in `FileStreamResultExecutor.cs`:

https://github.com/aspnet/Mvc/blob/aeab6e295efdc0b1d89c375b62a84fe3b6ba969e/src/Microsoft.AspNetCore.Mvc.Core/Internal/FileStreamResultExecutor.cs#L45-L48

https://github.com/aspnet/Mvc/blob/aeab6e295efdc0b1d89c375b62a84fe3b6ba969e/src/Microsoft.AspNetCore.Mvc.Core/Internal/FileStreamResultExecutor.cs#L65-L68

both fail to call `Dispose()`.

*Does the bug result in any actual functional issue, if so, what?*  

This can cause unmanaged resources to be leaked (e.g. open file handles preventing files from being manipulated).

### Minimal repro steps

See: https://github.com/Sumo-MBryant/FileStreamResultDisposeBug

In an MVC handler method:
1. Create a `Stream`
2. Create and return a `FileStreamResult`

```csharp
public IActionResult GetFileContent(string path)
{
    return new FileStreamResult(
        File.Open(path, FileMode.Open),
        "text/plain");
}
```

In client code:
1. Send an HTTP HEAD request
2. Attempt to delete file

```csharp
// Write some content to a file.
var content = "Hello";
var fileName = Path.GetRandomFileName();
File.WriteAllText(fileName, content);

// Fetch the header for the file.
using (var request = new HttpRequestMessage(
    HttpMethod.Head,
    $"test/test?path={WebUtility.UrlEncode(fileName)}"))
{
    using (var response = await client.SendAsync(request))
    {
    }
}

// Delete the file.
File.Delete(fileName);
```

### Expected result
The `FileStream` to be disposed on all executions of a `FileStreamResult`

### Actual result
The `FileStream` is not disposed if the content body isn't served.

### Further technical details

I think `FileStream.Dispose()` calls need to be added in both quoted pieces of source code. Version 1.1 didn't have these checks and always called `WriteFileAsync()`

Of course one could return a `FileResult` in the above example (and the whole example should never be written as such due to security considerations), but this affects all types of streams, many of which require `Dispose()` to be correctly called after their use.

### PR DESCRIPTION
**Pull request by kichalla**
**Note**: This issue still requires approval

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileStreamResultExecutor.cs [hunk 1]
```diff
@@ -47,6 +47,7 @@ public virtual Task ExecuteAsync(ActionContext context, FileStreamResult result)
 
             if (!serveBody)
             {
+                result.FileStream.Dispose();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileStreamResultExecutor.cs]: Should we put this inside a `using` block? In the event, any other piece of code in this type throws, we would return without disposing the stream.

Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileStreamResultExecutor.cs]: Sure, updated.

**Comment by kichalla** [comment]
@mkArtakMSFT We have following guards for preventing this
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/FileStreamResult.cs#L44
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/FileStreamResult.cs#L60

**Comment by kichalla** [comment]
https://github.com/aspnet/Mvc/commit/b56436b5c607e08ada5fccabfc71c475f8fbbbea

**Pull request closed without merging**